extern/recastnavigation/Detour/Source/DetourNavMesh.cpp

```diff
@@ -240,7 +240,7 @@ dtStatus dtNavMesh::init(const dtNavMeshParams* params)
 	m_posLookup = (dtMeshTile**)dtAlloc(sizeof(dtMeshTile*)*m_tileLutSize, DT_ALLOC_PERM);
 	if (!m_posLookup)
 		return DT_FAILURE | DT_OUT_OF_MEMORY;
-	memset(m_tiles, 0, sizeof(dtMeshTile)*m_maxTiles);
+	memset(static_cast<void*>(m_tiles), 0, sizeof(dtMeshTile)*m_maxTiles);
```

Comment:
It's not a false-positive when it's true. [It's undefined behavior to use `memset` on a non-trivially copyable type](https://en.cppreference.com/w/cpp/string/byte/memset), and [declaring non-defaulted and non-deleted copy methods makes it not trivially copyable](https://en.cppreference.com/w/cpp/named_req/TriviallyCopyable)). You don't know what the compiler may do to non-trivial structs, now or in the future, that can break this call.

Reply:
This is also 'extern' code. If there is a proper fix then we need to upstream it.

Reply:
According to Recast developers, declaring copy methods without implementation is a workaround for compilers, which do not support C++11:

> The structure being cleared is just a POD type with deleted copy-constructor and assignment operator.
It appears GCC8 does not recognize the old-style idiom used for deleting auto-generated members (where you just declare them privately without an implementation). We use this approach to support older compilers

Reply:
> According to Recast developers, declaring copy methods without implementation is a workaround for compilers, which do not support C++11

Yes, but all the compiler sees is the copy method declarations in private scope, and needs to assume they could be defined elsewhere, even if they ultimately never are. It's a useful pre-C++11 trick to prevent the copy operators from being used, but it makes it not trivially copyable as a consequence (a fact the compiler can leverage however it wants, including ways that make memset invalid on it).

> It appears GCC8 does not recognize the old-style idiom used for deleting auto-generated members (where you just declare them privately without an implementation).

It behaves exactly as it always has. It doesn't truly delete them, it just makes them not callable from outside of the class (and causes a linker error if it manages to do that). But the compiler can't know there's no implementation when parsing the class, since the methods can be in another source file.

Reply:
Are there "proper" fixes for this issue?

Reply:
Adding a `clear()` method to `dtMeshTile` (which sets all the members to 0/null), and calling it like:
```
std::for_each(m_tiles, m_tiles+m_maxTiles, std::mem_fn(&dtMeshTile::clear));
```
should work. Using proper construction and destruction would be better, but that would be more invasive changes.

Reply:
IIRC, the `std::mem_fn` is a C++11 thing, so it will not work. We need a fix which will work on all C++ versions. Unfortunately, developers tends to just disable the `-Wclass-memaccess`, so I did not find good alternatives so far.

Reply:
Pre-C++11 has `std::mem_fun`.

Reply:
> Pre-C++11 has std::mem_fun.

Which is deprecated in C++11 and removed in C++17, so there will be if-defs, which Recast developers will not accept. Weird.

Reply:
How about something like that:
```
    for (int i = 0; i < m_maxTiles; ++i)
	{
        m_tiles[i].clear();
    }
```

and:
```
    void clear()
    {
        salt = 0;
        linksFreeList = 0;
        header = NULL;
        polys = NULL;
        verts = NULL;
        links = NULL;
        detailMeshes = NULL;
        detailVerts = NULL;
        detailTris = NULL;
        bvTree = NULL;
        offMeshCons = NULL;
        data = NULL;
        dataSize = 0;
        flags = 0;
        next = NULL;
    }
```

Reply:
An another approach:
```
void* mem = dtAlloc(sizeof(dtMeshTile)*m_maxTiles, DT_ALLOC_PERM);
memset(mem, 0, sizeof(dtMeshTile)*m_maxTiles);
m_tiles = (dtMeshTile*)mem;
```

Reply:
@elsid, what do you think about this issue?

Reply:
> ```
> void* mem = dtAlloc(sizeof(dtMeshTile)*m_maxTiles, DT_ALLOC_PERM);
> m_tiles = (dtMeshTile*)mem;
> memset(mem, 0, sizeof(dtMeshTile)*m_maxTiles);
> ```

May as well just go the extra step:
```
void* mem = dtAlloc(sizeof(dtMeshTile)*m_maxTiles, DT_ALLOC_PERM);
memset(mem, 0, sizeof(dtMeshTile)*m_maxTiles);
m_tiles = ::new(mem) dtMeshTile[m_maxTiles];
```
That's the proper C++ way to create a new array in user-provided storage. To be extra proper, it should also run the destructor on the array before freeing:
```
for(int i = 0;i < m_maxTiles;++i)
    m_tiles[i].~dtMeshTile();
dtFree(m_tiles);
```

Reply:
```
void* mem = dtAlloc(sizeof(dtMeshTile)*m_maxTiles, DT_ALLOC_PERM);
memset(mem, 0, sizeof(dtMeshTile)*m_maxTiles);
m_tiles = ::new(mem) dtMeshTile[m_maxTiles];
```

Does not compile:
```
no matching function for call to «dtMeshTile::dtMeshTile()»
```

Reply:
```
for(int i = 0;i < m_maxTiles;++i)
    m_tiles[i].~dtMeshTile();
dtFree(m_tiles);
```

I suppose it is handled in this loop:
```
	for (int i = 0; i < m_maxTiles; ++i)
	{
		if (m_tiles[i].flags & DT_TILE_FREE_DATA)
		{
			dtFree(m_tiles[i].data);
			m_tiles[i].data = 0;
			m_tiles[i].dataSize = 0;
		}
	}
```

Reply:
You need to define a default/empty constructor since the copy constructor inhibits the automatic one.

Reply:
> I suppose it is handled in this loop

Technically, that if check and the extra `dtFree` call should be in `dtMeshTile`'s destructor. i.e.
```
dtMeshTile::~dtMeshTile()
{
    if(flags & DT_TILE_FREE_DATA)
        dtFree(data);
}

...

for(int i = 0;i < m_maxTiles;++i)
    m_tiles[i].~dtMeshTile();
dtFree(m_tiles);
```

Reply:
Are you sure the `::new(mem)` does not allocate new memory and uses the `mem` instead?

Reply:
> Are you sure the `::new(mem)` does not allocate a new memory and uses the mem instead?

I'm positive, that's called placement new. It constructs the objects in-place and returns a pointer to the correct type.

Reply:
I uploaded a new commit, despite I am still not convinced that touching a destructior is a good idea - suggested changed do not fit the `dtMeshTile` usage by R&D itself.

Reply:
First I'd try to suggest to remove private copy constructor and assignment operator. This type is a POD by intention. Actually to copy a POD is a correct behaviour since there is no destructor and user controls when to release any locked resources by an object of such type. They were added by https://github.com/recastnavigation/recastnavigation/commit/b7a83e0afbed5643600bad2b93febbad323333d1 that doesn't explicitly mentioned this in the commit message. Not sure is there any reasoning behind. IMO saying "it's a POD but we don't allow to copy it" is not consistent with C++.

If this will not be accepted, then I agree with kcat.

I've tried somehow to break code using such UB as memset on an non trivially copyable object. But didn't get any different results in assembly by different compilers. Probably compilers don't pay attention for it.

Reply:
So a suggested approach then:
1. Describe the issue to R&D devs (they disabled the `-Wclass-memaccess` via `premake` so far) and link this discussion, so they can make decision themselves - **done**.
2. Remove these private methods from our local copy of R&D, so we will silence this warning in OpenMW while upstream R&D does not have the fix.

Reply:
> I've tried somehow to break code using such UB as memset on an non trivially copyable object. But didn't get any different results in assembly by different compilers. Probably compilers don't pay attention for it.

The problem is ultimately that the language allows a compiler to do unorthodox things with non-trivial types, and the compiler is warning you to not do it. A compiler may not right now do anything that could break with memset, but it could at any time in the future without further warning. Disabling the warning would also hide legitimate problems, like if someone adds a `std::string` to a previously trivial type that gets used with `memset` or `memcpy`.

C++ libraries really have no reason to cater to pre-C++11 compilers, unless it's targeting old programs that aren't updated anymore and need an ABI of a particular compiler version. C++11 is over 8 years old, and has way too many features to improve the readability, safety, and maintenance of code. I might even argue that C++11 is getting outdated, but C++14 doesn't really add too much over it and I'd agree C++17 is still a bit too new to require. But as compilers improve, they're going to look at doing more and more things that can help code optimizations and tooling, so clever tricks you could get away with before aren't always going to be the ideal solution.

Reply:
C++14 is just a bug-fix on top of C++11, for anyone moving from 03, they should step right into 14.